src/compiler/checker.ts

```diff
@@ -6361,31 +6365,25 @@ namespace ts {
                 if (left.expression.kind !== SyntaxKind.Identifier || getResolvedSymbol(<Identifier>left.expression) !== symbol) {
                     return type;
                 }
-                let typeInfo = primitiveTypeInfo[right.text];
                 if (expr.operatorToken.kind === SyntaxKind.ExclamationEqualsEqualsToken) {
                     assumeTrue = !assumeTrue;
                 }
+                let typeInfo = primitiveTypeInfo[right.text];
+                // If the type to be narrowed is any and we're affirmatively checking against a primitive, return the primitive
+                if (!!(type.flags & TypeFlags.Any) && typeInfo && assumeTrue) {
+                    return typeInfo.type;
+                }
+                // At this point we can bail if it's not a union
+                if (!(type.flags & TypeFlags.Union)) {
+                    return type;
+                }
+                let flags = typeInfo ? typeInfo.flags : (assumeTrue = !assumeTrue, TypeFlags.NumberLike | TypeFlags.StringLike | TypeFlags.ESSymbol | TypeFlags.Boolean);

```

Comment:
PLEASE use `if/else`. Javascript is an imperative language, so people will miss `assumeTrue = !assumeTrue` in the middle of an expression (and middle of line).